src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/ReflectionXmlFormatReader.cs

```diff
@@ -449,7 +443,102 @@ private object ReflectionReadCollectionCore(CollectionDataContract collectionCon
+            {
+                return (resultCollection, collectionItem, index) =>
+                {
+                    resultCollection = XmlObjectSerializerReadContext.EnsureArraySize((T[])resultCollection, index);
+                    ((T[])resultCollection)[index] = (T)collectionItem;
+                    return resultCollection;
+                };
+            }
+            else if (collectionContract.Kind == CollectionKind.GenericDictionary || collectionContract.Kind == CollectionKind.Dictionary)
+            {
+                Type keyType = collectionContract.ItemType.GenericTypeArguments[0];
+                Type valueType = collectionContract.ItemType.GenericTypeArguments[1];
+                Func<object, object> objectToKeyValuePairGetKey = (Func<object, object>)s_objectToKeyValuePairGetKey.MakeGenericMethod(keyType, valueType).CreateDelegate(typeof(Func<object, object>));
+                Func<object, object> objectToKeyValuePairGetValue = (Func<object, object>)s_objectToKeyValuePairGetValue.MakeGenericMethod(keyType, valueType).CreateDelegate(typeof(Func<object, object>));
+
+                return (resultCollection, collectionItem, index) =>
+                {
+                    object key = objectToKeyValuePairGetKey(collectionItem);
+                    object value = objectToKeyValuePairGetValue(collectionItem);
+
+                    IDictionary dict = (IDictionary)resultCollection;
+                    dict.Add(key, value);
+                    return resultCollection;
+                };
+            }
+            else
+            {
+                Type collectionType = resultCollectionObject.GetType();
+                Type genericCollectionType = typeof(ICollection<T>);
+                Type typeIList = typeof(IList);
+                if (genericCollectionType.IsAssignableFrom(collectionType))
+                {
+                    return (resultCollection, collectionItem, index) =>
+                    {
+                        ((ICollection<T>)resultCollection).Add((T)collectionItem);
+                        return resultCollection;
+                    };
+                }
+                else if (typeIList.IsAssignableFrom(collectionType))
+                {
+                    return (resultCollection, collectionItem, index) =>
+                    {
+                        ((IList)resultCollection).Add(collectionItem);
+                        return resultCollection;
+                    };
+                }
+                else
+                {
+                    MethodInfo addMethod = collectionType.GetMethod("Add");
+                    if (addMethod == null)
+                    {
+                        throw System.Runtime.Serialization.DiagnosticUtility.ExceptionUtility.ThrowHelperError(new InvalidDataContractException(SR.Format(SR.CollectionMustHaveAddMethod, DataContract.GetClrTypeFullName(collectionContract.UnderlyingType))));
```

Comment:
Nit I just noticed this throw uses the ExceptionUtility, but the other places you throw are just "new InvalidOperationException()".  Should you be using the ExceptionUtility in those other places too?